File: src/aleph/dtos/shaping.py
"""Shaping API request/response DTOs (AL-320, Phase 2B TDD §6).

The wire contract for the shaping surface: the conversation read, the streamed
send, and the one new stream event a shaping reply can carry — ``proposal``.
Phase 2A's tutor contract (``dtos/tutor.py``) is **reused, not respelled**:
``TutorMessageStr`` bounds the learner's message, and ``delta`` / ``done`` /
``error`` ride the exact shapes the in-lesson rail already parses (§5.4 is
"Phase 2 §5.4 verbatim, plus one named event"). Only what genuinely differs
lives here.

**What differs, and only this.**

* **No lesson.** A shaping turn is about the path as a whole (PRD §5.1), so
  :class:`ShapingMessageDTO` carries no ``lesson_id`` and no ``lesson_title``.
  The columns exist and are ``NULL`` on these rows (migration ``0006``); a wire
  field that is always ``null`` would invite a client to render an empty lesson
  divider for a turn that was never in a lesson.
* **A Proposal may ride on a tutor message**, exactly as a Tutor check rides on
  one in the in-lesson thread — same position in the shape, same
  non-null-only-on-tutor-rows rule.

**Two proposal shapes, and the split is the wire contract.**

:class:`ProposalPayloadDTO` is the **bare validated payload** —
``{operations, summary}``, TDD §4's fixed shape and nothing else. It is what
the ``proposal`` SSE event carries and what is persisted on the tutor message
row, and those two being the same object is deliberate: the card the rail draws
mid-stream and the card it re-renders from a later thread read are then the same
card by construction.

:class:`ProposalDTO` is that payload plus ``resolution`` — *pending*, *applied*,
*undone* or *superseded* — for the conversation read. Resolution is **derived,
never stored** (D3), by
:func:`~aleph.services.tutor_context.derive_proposal_resolutions`, the same
function the shaper's carried history uses; computing it twice is how the card
and the model start disagreeing about whether an edit already landed. It is
absent from the stream event because a proposal that has just been made is
pending by definition, and a field whose only possible value is a constant is
noise on the wire.

**AL-321 adds the write half**: :class:`ShapingConflictReason` (the coded ``409``
a card renders when Apply or Undo refuses), :class:`ChangeDTO` and
:class:`ChangeHistoryResponse` (the read-only **Change history**), and
:class:`ApplyProposalResponse` (the committed Change plus the refreshed path, so
ghost rows become real rows in one round trip). Nothing in the conversation
contract above changed to accommodate them — see :class:`ProposalDTO` for why the
card carries no ``change_id``.

**The operations are the agent's own models**, imported from
``agents/shaper.py`` rather than restated: the closed two-shape vocabulary (D1)
is validated by predicates that live there, and a second declaration of the same
payload here is precisely how a DTO and a validator drift apart. Their
docstrings and field constraints therefore *are* the wire documentation.
"""

from __future__ import annotations

from datetime import datetime  # noqa: TC003 - pydantic resolves annotations.
from enum import StrEnum
from uuid import UUID  # noqa: TC003 - pydantic resolves annotations.

from pydantic import BaseModel, ConfigDict

from aleph.agents.shaper import ShapingOperation  # noqa: TC001 - pydantic needs it
from aleph.dtos.paths import PathDetailResponse  # noqa: TC001 - pydantic needs it
from aleph.dtos.tutor import TutorMessageStr  # noqa: TC001 - pydantic needs it
from aleph.models import (
    MessageRole,
    MessageSource,
    PathChangeKind,
    PathChangeStatus,
)


class ProposalResolutionDTO(StrEnum):
    """Where a Proposal stands (CONTEXT.md; TDD §4 — derived, never stored).

    A ``StrEnum`` member *is* its wire value, and the values are exactly
    ``agents/shaper.py``'s ``ProposalResolution`` literals: the derivation
    returns those strings and this enum is how they reach the client's card
    state machine (TDD §8) with one spelling.
    """

    #: No change row references it and nothing has superseded it — the card
    #: still offers **Apply**.
    PENDING = "pending"
    #: A live ``path_changes`` row references it: the edit landed.
    APPLIED = "applied"
    #: Its change was undone; the path is back where it was.
    UNDONE = "undone"
    #: A later Proposal was applied first and this one no longer validates
    #: against live path state — the card explains why and offers "ask again".
    SUPERSEDED = "superseded"


class ProposalPayloadDTO(BaseModel):
    """The validated **Proposal** payload: what the edit *is* (TDD §4).

    ``{operations, summary}`` — the closed vocabulary of **Additions** and
    **Revisions** (D1) plus the plain-language line stating what the payload
    does. This is the ``proposal`` SSE event's data, verbatim, and it is what is
    stored on the tutor message row; nothing else is added on either rail.

    Persisting a Proposal is not applying one. Nothing in this payload changes a
    path until the learner taps **Apply** (D5) — consent is structural, and this
    object is the statement of the offer, not of a change.
    """

    operations: list[ShapingOperation]
    summary: str


class ProposalDTO(ProposalPayloadDTO):
    """A stored Proposal as the conversation read reports it (§6).

    The payload plus its **derived** ``resolution`` (D3). The card renders its
    state from this one field, so it is computed by the shared derivation rather
    than by anything local to this surface.

    **There is deliberately no ``change_id``** (AL-321). The card's states are
    §8's — pending / applying / applied ("view in path") / stale / undone — and
    none of them needs one: **Apply** answers with the whole
    :class:`ChangeDTO`, so a card that just applied holds the id it needs, and
    **Undo** is reached from the change-history sheet, which carries every id by
    definition. Adding it here would mean a nullable field that is ``None`` for
    the entire life of the ordinary case, plus a second derivation to keep in
    step with ``resolution`` — the exact drift D3 avoided by not storing a status
    column. If a card ever grows an undo affordance of its own, the additive move
    is to put the id in the same derivation that already computes the state, not
    beside it.
    """

    resolution: ProposalResolutionDTO


class ShapingMessageDTO(BaseModel):
    """One message in the **Shaping conversation** — learner or tutor (§6).

    No lesson fields, deliberately (module docstring): shaping is path-level.
    ``proposal`` is non-null only on a tutor message that made one. The learner
    row's ``source`` stays off the wire for 2A's reason — it is the §7 entry-mix
    datum, carried by the product events, and the rail renders a learner bubble
    the same either way.
    """

    id: UUID
    role: MessageRole
    content: str
    proposal: ProposalDTO | None
    created_at: datetime


class ShapingConversationResponse(BaseModel):
    """``GET /api/v1/paths/{id}/shaping/conversation`` body — the whole thread.

    Object-wrapped (never a bare array) like every other list payload. Messages
    are in ``position`` order, oldest first.

    A path with no shaping conversation yet answers ``200`` with an empty list,
    not ``404``: the row is created lazily on the first completed turn (§5.5),
    so an untouched path and a cleared one read identically — which is exactly
    what "new conversation" should leave behind. **Unpaginated this phase**, the
    same accepted risk as the in-lesson thread (TDD §14).
    """

    messages: list[ShapingMessageDTO]


class ShapingConflictReason(StrEnum):
    """Why a ``409`` refused an **Apply** or an **Undo** (AL-321, TDD §5.8).

    Every one of these is an ordinary ``409 conflict`` in the shared envelope;
    this is the ``details.reason`` beside it, and it exists because §5.8 makes
    the stale path **first-class UX** rather than an error corner: the card has
    to say *which* thing changed and offer the matching affordance ("ask again",
    "retry in a moment", "this is permanent now"). A single opaque
    ``conflict`` would leave the rail guessing from prose.

    They partition into five groups the card treats differently:

    * **Nothing to do** — ``already_applied`` / ``already_undone`` /
      ``not_applied``. Idempotent-friendly: the learner (or a double tap) asked
      for a state the path is already in.
    * **Stale, so ask again** — ``path_cap_reached``, ``insert_position_taken``,
      ``revision_target_engaged``, ``title_conflict``, ``positions_shifted``,
      ``invalid_proposal``. The Proposal was valid when drafted and is not any
      more (D5); re-asking is the way forward, retrying is not.
    * **Retry in a moment** — ``target_generating``. A prefetch holds the claim;
      nothing is wrong and the same tap will work shortly.
    * **Undo something else first** — ``not_latest``. Nothing is wrong with this
      Change either; it is simply not the one on top of the stack.
    * **Closed for good** — ``engaged``. The learner met the content, so the
      Change is permanent history (PRD §5.5) and the UI says so plainly rather
      than hiding the button.
    """

    #: A live change row already references this Proposal.
    ALREADY_APPLIED = "already_applied"
    #: This Proposal was applied and undone; it is spent (ask again to redo it).
    ALREADY_UNDONE = "already_undone"
    #: Undo of a Change that is not ``applied`` (already undone).
    NOT_APPLIED = "not_applied"
    #: Undo of a live Change a *later* live Change was applied on top of. Undo
    #: is last-in-first-out: reverse the later one first.
    NOT_LATEST = "not_latest"
    #: The path no longer has room for the additions (``MAX_LESSONS_PER_PATH``).
    PATH_CAP_REACHED = "path_cap_reached"
    #: The insertion point is now before the learner's first non-engaged
    #: position, or past the end of the path.
    INSERT_POSITION_TAKEN = "insert_position_taken"
    #: A Revision names a lesson the learner has started since — or one that is
    #: no longer on this path.
    REVISION_TARGET_ENGAGED = "revision_target_engaged"
    #: A proposed title now collides with one already on the path.
    TITLE_CONFLICT = "title_conflict"
    #: An earlier Change moved the slot this Proposal's positions named.
    POSITIONS_SHIFTED = "positions_shifted"
    #: The payload no longer satisfies the shared predicates for some other
    #: reason (the catch-all, so the set stays closed as the rulebook grows).
    INVALID_PROPOSAL = "invalid_proposal"
    #: A revision target is generating right now — retryable in a moment.
    TARGET_GENERATING = "target_generating"
    #: The learner has engaged with what the Change created or revised, so undo
    #: is closed (D2 — the server is the enforcer, not the disabled button).
    ENGAGED = "engaged"


class ChangeDTO(BaseModel):
    """One applied **Change** as the history sheet reports it (§6).

    Plain-language summary, the edit shape(s), the status, the timestamps —
    TDD §6's list and nothing more. It is a **record, not a second edit
    surface** (PRD §5.5), so there is no payload here: the operations already
    rendered as a card in the thread, and the history is what happened, in the
    learner's own summary.

    ``kinds`` is plural because one Apply may carry both shapes: a Proposal's
    ``operations`` are a list of Additions *and* Revisions (D1), and they land
    as one Change because a Change is the unit of Apply **and** of Undo
    (CONTEXT.md) — undoing half of what the learner consented to as one edit
    would leave the path in a shape nobody proposed. The row's own ``kind``
    column names the change's dominant shape; this field is derived from the
    payload and is what the sheet renders.

    ``undone_at`` is ``None`` for a live Change. Whether undo is still *open* is
    deliberately **not** here: it is the D2 engagement re-check, run at undo
    time, because the learner can engage at any moment and a flag computed at
    read time would be stale before it was rendered. The ``409 engaged`` is the
    enforcer; the client's own knowledge of the path is the convenience.
    """

    id: UUID
    summary: str
    kinds: list[PathChangeKind]
    status: PathChangeStatus
    applied_at: datetime
    undone_at: datetime | None


class ChangeHistoryResponse(BaseModel):
    """``GET /api/v1/paths/{id}/changes`` body — the whole **Change history**.

    Object-wrapped (never a bare array) like every other list payload, and
    **newest first**: the history reads in the order the learner made it happen.
    Undone Changes are included — undo is a status, not a delete, and the
    history is the record of what happened (CONTEXT.md).

    It survives "new conversation" (PRD §5.8): the rows hang off the path, not
    off the thread, so this list is answered identically before and after a
    learner clears their shaping conversation.
    """

    changes: list[ChangeDTO]


class ApplyProposalResponse(BaseModel):
    """``POST /api/v1/messages/{id}/apply-proposal`` body (§5.6 step 4).

    The Change that was just committed **and the refreshed path**, in one round
    trip, so the rail can swap its **ghost rows** for real ones without a second
    request and without guessing what landed. ``path`` is exactly what
    ``GET /paths/{id}`` returns — the same query Phase 1's polling populates
    (TDD §8) — so the client drops it straight into the cache it already has.

    A Change is *applied* when the structure lands, not when generation finishes
    (PRD §5.7): the added rows in ``path`` are ``ungenerated`` and a revised one
    has gone back to ``ungenerated``, and Phase 1's untouched pipeline takes it
    from there.
    """

    change: ChangeDTO
    path: PathDetailResponse


class SendShapingMessageRequest(BaseModel):
    """``POST /api/v1/paths/{id}/shaping/conversation/messages`` body (§5.4).

    ``content`` reuses 2A's :data:`~aleph.dtos.tutor.TutorMessageStr` — stripped,
    non-empty, ≤ 2000 characters, rejected as a ``422`` *before* the stream opens
    — because it is the same thing: one learner message typed into one rail.

    There is no ``lesson_id``: a shaping turn is about the path as a whole
    (PRD §5.1). ``source`` records how the learner entered it (the four §5.3
    suggestions send as ``suggestion``), the §7 entry-mix datum.

    ``model`` is the **admin** per-message model override (TDD §5.3/D10),
    enforced exactly as 2A's: ``403`` for a non-admin (checked before the
    allowlist), ``422`` off ``MODEL_ALLOWLIST``, resolved per request and
    **persisted nowhere**. It binds the shaper slot (AL-301 put ``MODEL_SHAPER``
    behind the same allowlist), so an A/B of proposal quality costs no deploy.
    """

    # ``model`` starts with the ``model_`` prefix pydantic protects by default;
    # the picker's wire contract fixes the name (parity with ``MODEL_SHAPER``
    # and with 2A's ``SendMessageRequest``), so opt out of the protected
    # namespace rather than rename.
    model_config = ConfigDict(protected_namespaces=())

    content: TutorMessageStr
    source: MessageSource = MessageSource.TYPED
    model: str | None = None

File: src/aleph/web/frontend/tests/e2e/fixtures/shaping.ts
// Shared vocabulary for the shaping journeys (AL-331): opening the shaping rail
// on the path view, asking for an edit, and acting on the Proposal that comes
// back — plus the facts about the deterministic stub the specs assert against.
//
// It sits beside `tutor.ts` rather than inside it, for that file's own reason:
// the two rails are separate surfaces with separate testids and separate
// threads (PRD §5.8), and a helper that reached across them would be the first
// thing to break W21's "the in-lesson tutor stays bit-identical".
//
// Three rules it encodes:
//
//  1. **Nothing happens until Apply.** Every helper below that changes the path
//     goes through the card's own button. A Proposal on screen has changed
//     nothing, and `ghostRows` is how a spec says so.
//  2. **Sentinels live in the question**, stateless, stripped from the reply —
//     the same discipline the tutor's sentinels follow (`services/stub_model.py`).
//  3. **Structure, never wording.** The stub's prose is deterministic but its
//     phrasing is an implementation detail; what a spec asserts is the shape of
//     the payload and the state of the card.

import { type Locator, type Page, expect } from "@playwright/test";
import { ACTION_TIMEOUT, GENERATION_TIMEOUT } from "./journey";

/**
 * Sentinels the stub shaper reads out of the **question** text
 * (`services/stub_model.py`). Stripped from the reply, and stateless — each one
 * fires on every send that carries it.
 */
export const FORCE_PROPOSAL_ADD = "[force-proposal-add]";
export const FORCE_PROPOSAL_REVISE = "[force-proposal-revise]";
export const FORCE_SHAPING_DECLINE = "[force-shaping-decline]";
export const FORCE_SHAPING_FAILURE = "[force-shaping-failure]";

/**
 * How the stub titles what it proposes (`build_stub_addition_proposal` /
 * `build_stub_revision_proposal`). Restated here because a Playwright spec
 * cannot import Python; the Python definitions are the documented source. They
 * exist so an added or revised lesson is recognizable in a rail full of
 * generated ones — which is exactly what "the ghosts became real rows" needs.
 */
export const ADDED_LESSON_TITLE_PREFIX = "Added on request:";
export const REVISED_LESSON_TITLE_PREFIX = "Revised on request:";

/**
 * `REVISED_PASSAGE_MARKER` (`services/stub_model.py`), verbatim — W18's
 * assertion target.
 *
 * It is what closes the revision loop end to end in a browser: the sentinel puts
 * `SHAPING_REVISION_INSTRUCTION` on the Proposal, apply writes it to the
 * lesson's `revision_instruction`, the Phase 1 lesson prompt carries it in its
 * revision block (D7), and the stub — recognizing *its own* instruction — plants
 * this sentence in the regenerated Read passage. Seeing it on screen therefore
 * proves the instruction travelled the whole chain, which no wording check on
 * the reply could.
 */
export const REVISED_PASSAGE_MARKER = "This passage was regenerated from a learner's revision.";

/** The sentinel Addition's size — `_ADDITION_LESSON_COUNT` in the stub. */
export const ADDITION_LESSON_COUNT = 2;

/**
 * `SHAPING_DECLINED_EDIT_REPLY` (`services/stub_model.py`), verbatim — W20's
 * assertion target, and **Markdown source**, not rendered text.
 *
 * Asserted in full rather than by keyword because the **declined edit** is a
 * product promise (PRD §5.7): it names what shaping can do, and it must read as
 * neither a failure nor a safety refusal. A drift in it should be seen, not
 * absorbed — the same reason W15 pins the refusal copy.
 *
 * Compare it against the *wire* (`fetchShapingConversation`): the rail renders
 * this through `markdown.tsx`, so the `**add**` emphasis is a `<strong>` on
 * screen and the asterisks are gone from `innerText`.
 */
export const SHAPING_DECLINED_EDIT_REPLY =
  "That is not one of the changes I can make to a path. What shaping can do is " +
  "**add** lessons — on their own or grouped as a new unit — anywhere you have " +
  "not started yet, and **revise** a lesson you have not started yet so it " +
  "lands differently. What it cannot do is remove or reorder lessons, change " +
  "work you have already engaged with, or touch your progress. Your path is " +
  "exactly as it was. Tell me what you were hoping that change would get you " +
  "and there is a good chance an addition or a revision gets you there.";

/**
 * The rail's failure copy for a shaping turn that fell over — `_FAILURE_COPY`
 * for `upstream_error` in `services/tutor.py`.
 *
 * The *tutor's* module, and deliberately so: TDD §5.8 is a delta over Phase 2
 * §5.6, so both rails report a failed turn with one set of words. It is restated
 * here rather than imported from `tutor.ts` because what this constant means to
 * a shaping spec is "not a **declined edit**" — the two are different answers to
 * different situations, and W20 is where that distinction is asserted.
 */
export const SHAPING_UPSTREAM_FAILURE_COPY =
  "The tutor couldn't finish that answer. Nothing was saved — ask again when you're ready.";

/**
 * The server's `409 engaged` sentence, verbatim (`services/shaping.py`) — what
 * the Change history says when the undo window has closed for good.
 *
 * The client never pre-disables undo for engagement (it cannot derive it, and it
 * can change between the list rendering and the tap), so this copy is reached by
 * *tapping* — which is exactly the promise W19 asserts: say plainly that the
 * window is closed rather than hide the button.
 */
export const UNDO_ENGAGED_COPY =
  "you have already started one of the lessons this change made, so it is now " +
  "part of your path's history";

// --- Opening the shaping rail -------------------------------------------------

/** Open the shaping rail from its floating mark — the phone's way in (PRD §5.1). */
export async function openShapingRail(page: Page): Promise<void> {
  await page.getByTestId("shaping-rail-mark").click();
  await expect(page.getByTestId("shaping-rail")).toBeVisible({ timeout: ACTION_TIMEOUT });
}

/** Collapse the shaping rail; the mark comes back in its place. */
export async function closeShapingRail(page: Page): Promise<void> {
  await page.getByTestId("shaping-rail-collapse").click();
  await expect(page.getByTestId("shaping-rail")).toHaveCount(0);
  await expect(page.getByTestId("shaping-rail-mark")).toBeVisible();
}

// --- The thread ---------------------------------------------------------------

/** The tutor's replies in the shaping thread. */
export function shapingReplies(page: Page): Locator {
  return page.locator('[data-testid="shaping-rail-message"][data-role="tutor"]');
}

/** Every proposal card in the thread, oldest first. */
export function proposalCards(page: Page): Locator {
  return page.getByTestId("shaping-rail-proposal");
}

/**
 * Ask the shaping tutor something and wait for the whole turn to settle: the
 * reply appended, the live bubble gone, the composer open again.
 *
 * `GENERATION_TIMEOUT` for the same reason `askTutor` uses it — a turn is a
 * model call plus a persist, which on a loaded runner is a generation's worth of
 * wall time even though the stub answers instantly.
 */
export async function askShaper(page: Page, question: string): Promise<void> {
  const before = await shapingReplies(page).count();
  await page.getByTestId("shaping-rail-input").fill(question);
  await page.getByTestId("shaping-rail-send").click();
  await expect(shapingReplies(page)).toHaveCount(before + 1, { timeout: GENERATION_TIMEOUT });
  await expect(page.getByTestId("shaping-rail-streaming")).toHaveCount(0);
  await expect(page.getByTestId("shaping-rail-send")).toBeVisible();
}

/**
 * Every Proposal ask gets a unique subject. The stub seeds the titles it
 * proposes from the question text, so two identical asks in one journey produce
 * identical titles — and once the first is applied, the second proposal fails
 * the distinct-titles predicate server-side and never reaches the rail.
 * Real learners hit this only by asking for the same thing twice verbatim;
 * the counter keeps the harness off that edge.
 *
 * One counter for both shapes, because the collision is about *titles* and both
 * builders draw theirs from the same question text.
 */
let proposalAskCounter = 0;

/** Ask for an Addition and wait for the card it produces to be pending. */
export async function askForAddition(page: Page): Promise<Locator> {
  proposalAskCounter += 1;
  await askShaper(page, `Add practice round ${proposalAskCounter} on this ${FORCE_PROPOSAL_ADD}`);
  return pendingCard(page);
}

/** Ask for a Revision and wait for the card it produces to be pending. */
export async function askForRevision(page: Page): Promise<Locator> {
  proposalAskCounter += 1;
  await askShaper(
    page,
    `Make my next lesson simpler, take ${proposalAskCounter} ${FORCE_PROPOSAL_REVISE}`,
  );
  return pendingCard(page);
}

/** The newest card in the thread, once it is offering itself. */
async function pendingCard(page: Page): Promise<Locator> {
  const card = proposalCards(page).last();
  await expect(card).toHaveAttribute("data-state", "pending", { timeout: ACTION_TIMEOUT });
  return card;
}

// --- Ghost rows ---------------------------------------------------------------

/** The **ghost rows** the path rail is previewing right now (CONTEXT.md). */
export function ghostRows(page: Page): Locator {
  return page.getByTestId("path-rail-ghost");
}

/** The rows a pending **Revision** has marked "will be revised". */
export function revisingRows(page: Page): Locator {
  return page.getByTestId("path-rail-revising");
}

// --- Apply, Undo, the Change history -----------------------------------------

/**
 * Tap **Apply** on a card and wait for the Change to land — the card applied and
 * the rail's ghosts gone, because the response's `path` swapped them for real
 * rows in the same round trip (TDD §5.6).
 */
export async function applyProposal(page: Page, card: Locator): Promise<void> {
  await card.getByTestId("shaping-rail-proposal-apply").click();
  await expect(card).toHaveAttribute("data-state", "applied", { timeout: GENERATION_TIMEOUT });
  await expect(ghostRows(page)).toHaveCount(0);
}

/** `POST /messages/{id}/apply-proposal` — the one write into path structure. */
const APPLY_ROUTE = /\/api\/v1\/messages\/[^/]+\/apply-proposal$/;

/** One lesson as the apply response draws it (`path` is `GET /paths/{id}`). */
export interface AppliedLesson {
  id: string;
  title: string;
  generation_state: string;
}

/** The `200 {change, path}` body apply answers with (docs/api.md). */
export interface AppliedPayload {
  change: ChangePayload;
  path: { units: { lessons: AppliedLesson[] }[] };
}

/**
 * Apply, and hand back the body the server answered with.
 *
 * For the one claim the rendered rail cannot hold still for: **apply lands
 * structure, not content** (PRD §5.7). Added lessons arrive `ungenerated` and
 * ride Phase 1 from there — and because the same response kicks the prefetch
 * driver, they can be generated by the time any assertion reaches the DOM.
 * Reading the response is not a workaround for that race; it is the exact
 * statement of the claim, at the moment the claim is about.
 */
export async function applyProposalReadingResponse(
  page: Page,
  card: Locator,
): Promise<AppliedPayload> {
  const response = page.waitForResponse(
    (candidate) => APPLY_ROUTE.test(candidate.url()) && candidate.request().method() === "POST",
  );
  await applyProposal(page, card);
  return (await (await response).json()) as AppliedPayload;
}

/** Every lesson in an applied path payload, in path order. */
export function appliedLessons(payload: AppliedPayload): AppliedLesson[] {
  return payload.path.units.flatMap((unit) => unit.lessons);
}

/** Open the read-only **Change history** sheet from the rail's header. */
export async function openChangeHistory(page: Page): Promise<void> {
  await page.getByTestId("shaping-rail-change-history").click();
  await expect(page.getByTestId("shaping-rail-history")).toBeVisible({ timeout: ACTION_TIMEOUT });
}

/** Close the sheet and return to the conversation it covered. */
export async function closeChangeHistory(page: Page): Promise<void> {
  await page.getByTestId("shaping-rail-history-close").click();
  await expect(page.getByTestId("shaping-rail-history")).toHaveCount(0);
}

/** Every row of the Change history, newest first. */
export function changeRows(page: Page): Locator {
  return page.getByTestId("shaping-rail-history-change");
}

/** Undo the newest live Change from the sheet and wait for it to read undone. */
export async function undoNewestChange(page: Page): Promise<void> {
  const newest = changeRows(page).first();
  await newest.getByTestId("shaping-rail-history-undo").click();
  await expect(newest).toHaveAttribute("data-status", "undone", { timeout: GENERATION_TIMEOUT });
}

// --- Wire reads ---------------------------------------------------------------

/** A `GET` on the app's own API, asserted OK and parsed. */
async function fetchJson<T>(page: Page, apiPath: string): Promise<T> {
  const response = await page.request.get(`/api/v1${apiPath}`);
  expect(response.ok()).toBe(true);
  return response.json() as Promise<T>;
}

export interface ChangePayload {
  id: string;
  summary: string;
  kinds: string[];
  status: string;
  applied_at: string;
  undone_at: string | null;
}

/**
 * The Change history straight off `GET /paths/{id}/changes`. The DOM can only
 * speak for what was rendered; a spec asserting that the record *survived*
 * something has to speak for what was stored.
 */
export function fetchChanges(page: Page, pathId: string): Promise<{ changes: ChangePayload[] }> {
  return fetchJson(page, `/paths/${pathId}/changes`);
}

/** The shaping thread off the wire — the other thread, never the in-lesson one. */
export function fetchShapingConversation(
  page: Page,
  pathId: string,
): Promise<{ messages: { role: string; content: string; proposal: unknown }[] }> {
  return fetchJson(page, `/paths/${pathId}/shaping/conversation`);
}
